hop/io.py

```diff
@@ -461,18 +730,160 @@ def __init__(self, broker_addresses, topic, auth, **kwargs):
                 time to wait for a message to be sent to Kafka. If zero, sending
                 will never time out.
+            automatic_offload: If true, when a message is too large for the target
+                               topic, and if the broker declares a suitable endpoint,
+                               offload the message to the offload service, and send
+                               a place-holder 'external' message on the Kafka topic
+                               in its place.
+            topic_check_period: Period to wait before checking whether the settings
+                of a target topic have changed.
 
         :meta private:
         """
-        logger.info(f"connecting to kafka://{','.join(broker_addresses)}")
-        self._producer = producer.Producer(producer.ProducerConfig(
-            broker_urls=broker_addresses,
-            topic=topic,
-            auth=auth,
-            **kwargs,
-        ))
+        self.topics = {}
+        self.producers = {}
+        self.broker_addresses = broker_addresses
         self.auth = auth
-        logger.info(f"publishing to topic: {topic}")
+        self.automatic_offload = automatic_offload
+        self.topic_check_period = topic_check_period
+
+        self.producer_args = _filter_valid_args(adc_producer.ProducerConfig, kwargs)
+        self.producer_args["auth"] = self.auth
+        if isinstance(broker_addresses, str):
+            broker_addresses = [broker_addresses]
+        self.producer_args["broker_urls"] = self.broker_addresses
+
+        if len(topics) == 1:
+            self.default_topic = topics[0]
+        else:
+            self.default_topic = None
+
+        if len(topics) != 0:
+            logger.info(f"connecting to kafka://{','.join(broker_addresses)}")
+            settings = self._check_topic_settings(topics)
+            for topic, topic_settings in settings.items():
+                self._set_producer_for_topic(topic, topic_settings)
+            logger.info(f"publishing to topic(s): {topics}")
+
+    def _check_topic_settings(self, topics):
+        aconfig = adc_producer.ProducerConfig(broker_urls=self.broker_addresses,
+                                              auth=self.auth,
+                                              topic=None,
+                                              )
+        aclient = AdminClient(aconfig._to_confluent_kafka())
+        logger.debug(f"Fetching settings for topics: {topics}")
+        query = [ConfigResource(restype=ResourceType.TOPIC, name=topic) for topic in topics]
+        futures = aclient.describe_configs(query)
+        results = {}
+        for resource, future in futures.items():
+            try:
+                results[resource.name] = future.result()
+            except confluent_kafka.KafkaException as ke:
+                kerr = ke.args[0]
+                # it is a potentially common issue that a user has WRITE permission but not
```

Comment:
Why is this a potentially common issue? I would think it would be good to allow any user to get the max message size?

Reply:
I think it would too, but the designers of Kafka made these completely independent permissions, and e.g. our own AuthZ system did not allow querying topic properties by default along with write permission [until I fixed it last week](https://github.com/scimma/kafka-plugins/commit/3b52ee611f816e52716f849e15e43cd0ce9e83a0), so I expect this misconfiguration to exist on many brokers.